### ISSUE TITLE
Import errors when there both a .css and .scss file in the specified location.

### ISSUE THREAD

**Issue by dbatiste** [sass/node-sass#1222]
Given dependency path that has both:
```
foo.css
foo.scss
```
Attempting to import `foo.scss` in a dependent file:
```
@import 'foo';
```
results in an error:

{
  "message": "It's not clear which file to import for '@import \"foo\"'.\nCandidates:\n  foo.scss\n  foo.css\nPlease delete or rename all but one of these files.\n",
  "column": 1,
  "line": 1,
  "file": "application.scss",
  "status": 1
}

This worked prior to the recent update.  Is this change intentional, bug?  Explicitly specifying the extension in the `@import` statement resolves the issue for a particular case, however this is breaking for many components.

**Comment by mgreter** [comment]
This is intentional and equal to how ruby sass handles this case.

**Comment by austinsmorris** [comment]
I am also getting this issue when there is two .scss files, but one is proceeded with an underscore.  For example:

`@import "path/to/foo";`

And then in the `/path/to/` directory, I have both `foo.scss` and `_foo.scss`.

This was working prior to 3.4.0.

**Comment by xzyfer** [comment]
This is same behaviour as Ruby Sass. Node Sass is just an implementation of the Ruby Sass spec so must do everything they do as they do it. The old behaviour was a bug.

*****

Using Sass 3.4.19. Given this folder.

```
$ ls -l
total 32
-rw-r--r--  1 michael  staff  13 27 Oct 23:36 _foo.scss
-rw-r--r--  1 michael  staff  13 27 Oct 23:36 foo.css
-rw-r--r--  1 michael  staff  10 27 Oct 23:36 foo.scss
-rw-r--r--  1 michael  staff  26 27 Oct 23:37 index.scss
```
```
$ echo "@import 'foo'" | sass -s
Error: It's not clear which file to import for '@import "foo"'.
       Candidates:
         _foo.scss
         foo.scss
       Please delete or rename all but one of these files.
        on line 1 of standard input
  Use --trace for backtrace.
```

**Comment by mgreter** [comment]
This is the case because libsass allows embeddable CSS partials by default. If you enable this for ruby sass, you will get the same error too. Only improvement I would see is to add a config option to enable/disable this feature, as I have suggested before.

**Comment by xzyfer** [comment]
I agree, we should add the flag and have it disabled by default to match the Ruby behaviour.

**Comment by rankida** [comment]
This has caused us great issues in our project and have had to revert back to `3.3.0` in the mean time.

I would vote for the feature to be turned off/on, though I am surprised that the default is on. Indeed I am surprised that when given the option of `filename.scss` and `filename.css` the `.scss` doesn't simply win by default.

Must say, good work and thanks to the project team - node-sass is great!

**Comment by xzyfer** [comment]
I've created https://github.com/sass/libsass/issues/1656 to track this fix in LibSass.

**Comment by hkassam** [comment]
I would actually suggest that, since this is a breaking change with the way it was working before, it requires a major version bump (as this change breaks every project that references node-sass in their package.json as "^3.x.y").

**Comment by chriseppstein** [comment]
Sass only gives this error if there are sass files (.sass, .scss) with the same basename after removing the partial prefix. It is common practice for people to generate their css files into the same folder as their sass files (I wouldn't do this, but many do).

In terms of when node-sass should bump major versions I would argue that any time node-sass makes a breaking change to match ruby sass's behavior, this is simply a bug fix and does not warrant a major number bump.

**Comment by mgreter** [comment]
@chriseppstein FYI: sass behaves the same if the "option" `super.merge('css' => :scss)` is set ...

**Comment by hkassam** [comment]
@chriseppstein the problem with that is, anyone using node-sass has been using it a specific way and expects it to work that way (regardless of whether it works the same as ruby-sass or not).  Changing this behaviour in a breaking change means EVERY project utilizing this module will break.  If node-sass is using semantic versioning, breaking changes require a major version bump (http://semver.org/)

**Comment by chriseppstein** [comment]
@hkassam I know how semver works. But it will be a huge pain in the ass from a marketing and usability perspective if node-sass and ruby-sass are feature compatible and not using the same major/minor number for language syntax purposes. Also: as far as breaking changes go, this one is super easy to work around. 

@mgreter that is not a default setting for Ruby Sass.

**Comment by hkassam** [comment]
Ok, so I understand the headache of having node-sass and node-ruby being feature compatible being extremely desirable.  I'm just saying that this change was breaking and should have updated the major version so as not to break ALL of the projects that are utilizing node-sass.  An assumption that npm seems to use is that all node packages utilize semver.  I would suggest, as it doesn't seem node-sass uses semver (which is fine), that a comment is made in the readme to indicate this.

**Comment by dbatiste** [comment]
From what I understand, and given @xzyfer 's comment above, it seems that this should have been a deprecation warning.  This would also have given folks some time to make their includes more explicit before encountering build errors such as this.
 
When people use semver, they are trusting that breaking changes will not be introduced without an appropriate version change.  Semver provides benefits to both providers and consumers - one of those benefits is that the dev community shares the testing, however when consumers can't trust revisions and minor version bumps, then they start opting out and pegging to specific versions, and this makes me sad.  

It doesn't really matter how easy the work-around is.  I appreciate the desire to keep the versions the same - perhaps that's an indication they shouldn't be maintained independently.

**Comment by xzyfer** [comment]
This was a breaking change that was not caught by either our tests or LibSass.

We will fix LibSass to match the Ruby Sass behaviour https://github.com/sass/libsass/issues/1656#issuecomment-156525091

Sass 4.0 will introduce a module system which will give the user more flexibility with their imports.

**Comment by pleunv** [comment]
Would it not make more sense to default to the scss file when both a scss and css file are available, perhaps with an optional warning that a file conflict exists? This would allow you to use the new css @imports while simultaneously avoiding having to change most of your existing @import statements.

We have multiple themes importing a base scss file and several npm components in order to recompile it with extensions and updated variables. All of these imported files get compiled individually as well for use in projects that do not use Sass. Consequently all of these now throw errors if we update node-sass and would require us to change dozens of @import statements across multiple projects. Having a mix of @imports that contain extensions and others that do not is also pretty confusing and really consistent behavior.

**Comment by airtonix** [comment]
seriously, there is a vast and immense chasm between the number of people who care about the vanity of a version number and those who just don't want (or can't budget time) to fix projects (now broken) because of a vanity number.

Just use semver properly please.

**Comment by xzyfer** [comment]
@airtonix it appears you have not fully read the issue. This is a bug in a dependency that was not caught by their tests. Node Sass is powerless here. This is an uncommon edge case. The problem should be fixed the upcoming 3.5.0 beta. 

I have no idea what your complaint about vanity numbers is.

In the future please ask yourself whether your comment adds value to problem at hand. If you're looking for software free of bugs I wish you luck in your search.

**Comment by baamenabar** [comment]
What we ended up doing was:
Find and replace in the whole project using the following pattern.
`@import ('|")((?!compass).+)(?<!scss)('|")`
the replace with 
`@import$1$2.scss$3`
We have literally hundreds of imports, and at least 50 had this problem. 

I hope this helps someone looking for this error.

**Comment by mcropper** [comment]
So what's the fix for this? If I'm explicit in the @import another error (EISDIR) seems to arise with the regards to the mixins folder.

**Comment by d07RiV** [comment]
Doesn't this completely defeat the purpose of having an extension-less version of `@import`? Since an `.scss` file will usually be accompanied by a (generated) `.css` file, you will never be able to import it without explicitly specifying the extension. This sounds like a bug to me, especially since the error message says:

    Please delete or rename all but one of these files.

Sure.. I'll delete one, only to have it reappear next time I update the other.

**Comment by xzyfer** [comment]
>Since an .scss file will usually be accompanied by a (generated) .css file

Output to the input directory is not support. We suggest separating input and output i.e. `src` and `dist` - since only input files should be committed.

**Comment by nschonni** [comment]
@nex3 should this be added as an error to Ruby/Dart?

**Comment by xzyfer** [comment]
@nschonni this is a known bug in LibSass see https://github.com/sass/libsass/pull/754 for background. It's a feature/bug I've been trying to [back out a couple years][1] now. The consensus atm is to leave the broken behaviour in place until we have first class CSS import support via Sass 4 modules. 

[1]: https://github.com/sass/libsass/pull/1963

**Comment by nschonni** [comment]
@xzyfer sorry, I know about the upstream issue, I was mixing this up with the issue about input and output directories overlapping

**Comment by AldoMX** [comment]
Hello, the recommended way to use sass by [create-react-app](https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/template/README.md#adding-a-css-preprocessor-sass-less-etc) is to overlap source and destination directories, which eventually results in this issue.

I understand that keeping backwards-compatibility with ruby sass is your top priority, but please add a command line argument to override this behavior.
